src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -1458,6 +1457,39 @@ internal virtual bool TryApplyChanges(Solution newSolution, IProgress<CodeAnalys
             }
         }
 
+        private void ApplyDocumentsInfoChange(ImmutableArray<ProjectChanges> projectChanges)
+        {
+            using var _1 = PooledHashSet<DocumentId>.GetInstance(out var infoChangedDocumentIds);
+            using var _2 = PooledHashSet<Document>.GetInstance(out var infoChangedNewDocument);
+            foreach (var projectChange in projectChanges)
+            {
+                foreach (var docId in projectChange.GetChangedDocuments())
+                {
+                    var oldSolution = projectChange.OldProject.Solution;
+                    var oldDoc = projectChange.OldProject.GetRequiredDocument(docId);
+                    var newDoc = projectChange.NewProject.GetRequiredDocument(docId);
+                    if (oldDoc.HasInfoChanged(newDoc))
+                    {
+                        var linkedDocumentIds = oldSolution.GetRequiredDocument(docId).GetLinkedDocumentIds();
+                        // For linked documents, when info get changed (e.g. name/folder/filePath)
+                        // only apply one document changed because it will update the 'real' file, causing the other linked documents get changed.
+                        if (linkedDocumentIds.All(linkedDocId => !infoChangedDocumentIds.Contains(linkedDocId)))
+                        {
+                            infoChangedDocumentIds.Add(docId);
+                            infoChangedNewDocument.Add(newDoc);
+                        }
+                    }
+                }
+            }
```

Comment:
This seems like it could break things that should work.  Say, for example, the user performs a fix-all on code like:

```c#
#if DEBUG

// code that is active in some project

#else

// other code that is not.

#endif
```

with the normal batch fixer, this would work, right?  Or maybe it wouldn't?  I'm not exactly sure.

Reply:
I feel it won't be affected. This change would only affect things like file name, path change. For content change, each document would still be changed even they are linked.

Reply:
good ot know!  probably worth a test :)

Reply:
I imagine we might end up actually taking the ApplyChangedDocument path more than once, but we'll end up writing the same text each time -- the documents should have had merged texts. If we look under a debugger and see we are indeed doing that, that might be a nice potential win here to only process that once.

That said, I'm not sure how we exactly fit that in here -- we're just calling ApplyProjectChanges for both projects and we can't just pass additional parameters since that's a protected API. So either we have to break the API, or maybe pre-filter the ProjectChanges to remove the duplicate changes.

Reply:
So, it sounds like a future work? (e.g. we have to see how things work for text changes)